crates/eframe/src/native/run.rs

```diff
@@ -65,23 +28,53 @@ enum EventResult {
 
 trait WinitApp {
     fn is_focused(&self) -> bool;
-    fn integration(&self) -> &EpiIntegration;
-    fn window(&self) -> &winit::window::Window;
+    fn integration(&self) -> Option<&EpiIntegration>;
+    fn window(&self) -> Option<&winit::window::Window>;
     fn save_and_destroy(&mut self);
     fn paint(&mut self) -> EventResult;
-    fn on_event(&mut self, event: winit::event::Event<'_, RequestRepaintEvent>) -> EventResult;
+    fn on_event(
+        &mut self,
+        event_loop: &EventLoopWindowTarget<RequestRepaintEvent>,
+        event: winit::event::Event<'_, RequestRepaintEvent>,
+    ) -> EventResult;
+}
+
+#[allow(unused)]
+fn create_event_loop_builder(
+    native_options: &mut epi::NativeOptions,
+) -> EventLoopBuilder<RequestRepaintEvent> {
+    let mut event_loop_builder = winit::event_loop::EventLoopBuilder::with_user_event();
+
+    if let Some(hook) = std::mem::take(&mut native_options.event_loop_builder) {
+        hook(&mut event_loop_builder);
+    }
+
+    event_loop_builder
 }
 
 /// Access a thread-local event loop.
 ///
 /// We reuse the event-loop so we can support closing and opening an eframe window
 /// multiple times. This is just a limitation of winit.
-fn with_event_loop(f: impl FnOnce(&mut EventLoop<RequestRepaintEvent>)) {
+fn with_event_loop(
+    mut native_options: epi::NativeOptions,
+    f: impl FnOnce(&mut EventLoop<RequestRepaintEvent>, NativeOptions),
+) {
     use std::cell::RefCell;
-    thread_local!(static EVENT_LOOP: RefCell<EventLoop<RequestRepaintEvent>> = RefCell::new(winit::event_loop::EventLoopBuilder::with_user_event().build()));
+    thread_local!(static EVENT_LOOP: RefCell<Option<EventLoop<RequestRepaintEvent>>> = RefCell::new(None));
 
     EVENT_LOOP.with(|event_loop| {
-        f(&mut *event_loop.borrow_mut());
+        // Since we want to reference NativeOptions when creating the EventLoop we can't
+        // do that as part of the lazy thread local storage initialization and so we instead
+        // create the event loop lazily here
+
+        let mut event_loop = event_loop.borrow_mut();
+        if event_loop.is_none() {
+            event_loop.replace(create_event_loop_builder(&mut native_options).build());
+        }
+        if let Some(event_loop) = event_loop.as_mut() {
+            f(event_loop, native_options);
+        }
```

Comment:
This can be simplified by using [`Option::get_or_insert_with`](https://doc.rust-lang.org/stable/std/option/enum.Option.html#method.get_or_insert_with).

Reply:
:) I guess I have a blind spot for some combinator APIs - a vaguely recall you suggested something similar with my previous PR